File: 240320/최소가 되는 좌표 범위의 차/the-minimum-difference-in-coordinate-range.py
import sys
import bisect

N, D = map(int,sys.stdin.readline().split())
dots = {}
for _ in range(N):
    x, y = map(int,sys.stdin.readline().split())
    try:
        dots[y].append(x)
    except:
        dots[y] = [x]

index = list(sorted(dots.keys()))
if index[-1]-index[0] < D:
    print(-1)
    exit(0)

dist = 10**7
for left in range(len(index)):
    right = len(index)-1
    while left < right:
        if index[right] - index[left] >= D:
            for lx in dots[index[left]]:
                lidx = bisect.bisect_left(dots[index[right]],lx)
                ridx = bisect.bisect_right(dots[index[right]],lx)
                if lidx == ridx:
                    dist = min(dist,abs(dots[index[right]][0]-lx),abs(dots[index[right]][-1]-lx))
                else:
                    for idx in range(lidx,ridx):
                        dist = min(dist,abs(dots[index[right]][idx]-lx))
                
        right -= 1

print(dist)

'''
dots = [tuple(map(int,sys.stdin.readline().split())) for _ in range(N)]
dots.sort(key=lambda x: (x[1],x[0]))

if dots[-1][1] - dots[0][1] < D:
    print(-1)
    exit(0)

dist = 10**7

for left in range(N):
    right = N-1
    while left < right:
        if dots[right][1] - dots[left][1] >= D:
            dist = min(dist,abs(dots[left][0]-dots[right][0]))
            right -= 1
        else:
            break

print(dist)
'''

File: 240321/최소가 되는 좌표 범위의 차/the-minimum-difference-in-coordinate-range.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define min(x, y) (x) < (y) ? (x) : (y)

int dot_comp(const void* d1, const void* d2){
    int *a, *b;
    a = (int*)d1;
    b = (int*)d2;
    if (a[1] < b[1]) return -1;
    else return 1;
}

int main() {
    // 여기에 코드를 작성해주세요.
    int N, D;
    scanf("%d %d",&N,&D);
    int dots[N][2];

    for(int i = 0; i < N; i++) {
        scanf("%d %d",&dots[i][0],&dots[i][1]);
    }

    qsort(dots,N,sizeof(dots[0]),dot_comp);

    int dist = 10000000;
    int left, right;

    for(left = 0; left < N; left++) {
        right = N-1;
        while(left < right) {
            if (dots[right][1]-dots[left][1] >= D) {
                dist = min(dist,abs(dots[left][0]-dots[right][0]));
                right--;
            }
            else break;
        }
    }

    printf("%d\n",dist);

    return 0;
}